source/common/http/filter_manager.cc

```diff
@@ -98,11 +98,16 @@ void ActiveStreamFilterBase::commonContinue() {
 
   doMetadata();
 
+  // It is possible for trailers to be added during doData(). doData() itself handles continuation
+  // of trailers for the non-continuation case. Thus, we must keep track of whether we had
+  // trailers prior to calling doData(). If we do, then we continue them here, otherwise we rely
+  // on doData() to do so.
```

Comment:
lazy ask, can we have a similar problem adding trailers while processing headers / metadata or is it only a data problem?

Reply:
I'm traveling today/tomorrow and in all day meetings. I'm inclined to get this fix in prior to release and I can research your question separately? Or do you want to hold this until I investigate? I think this fix will cover the header case also, but I need to do some research.

Reply:
SGTM.  Bonus points for TODO or tracking bug but really I trust you're good for it :-P

Reply:
Also if we're adding close to the release maybe runtime guard just in case there's bugs?  Optional, because I'm happy to defer to your judgement here.

Reply:
I looked into this and currently we only "allow" Add*Trailers() to be called during decodeData()/encodeData() when end_stream is true. This is tracked by the following assertion:

```
// Trailers can only be added during the last data frame (i.e. end_stream = true).
ASSERT(state_.filter_call_state_ & FilterCallState::LastDataFrame);
```

Given that we don't support adding trailers in headers/metadata today, I think it's probably safe to move forward with this fix. I'm OK without the runtime guard given how targeted the fix is and that without it the code will crash or do something else very bad so I'm going to go ahead and merge.